Assets/Scripts/Edges/EdgeBuilder.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Assets.Scripts.Model;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Assets.Scripts.Edges
+{
+  public class EdgeBuilder : IEdgeBuilder
+  {
+    const float c_edgeThicknessMultiplier = 0.1f;
+
+    private readonly Func<GameObject, GameObject> _instantiateMethod;
+    private readonly MaterialCache _materialCache;
+    private readonly LineFactory _lineFactory;
+    private List<Vector3> _intermediatePoints;
+    private Vector3? _startingNode = null;
+    private Vector3? _endingNode = null;
+    private float? _rotation = null;
+    private GraphElementClass _class;
+    private GameObject _lineStart;
+    private GameObject _lineEnd;
+    private GameObject _label;
+
+    public EdgeBuilder(MaterialCache materialCache, Func<GameObject, GameObject> instantiateMethod)
+    {
+      _materialCache = materialCache;
+      _lineFactory = new LineFactory();
+      _instantiateMethod = instantiateMethod;
+    }
+
+    public IEdgeBuilder BetweenNodes(Node node1, Node node2)
+    {
+      _startingNode = node1.Position;
+      _endingNode = node2.Position;
+      return this;
+    }
+
+    public IEdgeBuilder OnOneNode(Node node)
+    {
+      _startingNode = node.Position;
+      _endingNode = node.Position;
+      return this;
+    }
+
+    public IEdgeBuilder Curved(float rotation)
+    {
+      _rotation = rotation;
+      return this;
+    }
+
+    public IEdgeBuilder WithLabel(Func<GameObject> labelFactory, string label, bool visibility)
```

Comment:
I would like labelProducer better or labelFactoryFunction, when I see "factory" I immediately think of an object

Reply:
Sure, will rename to labelFactoryMethod. 
It's pretty common to call delegates that create objects "factories" though I understand the confusion.